### ISSUE TITLE
Upload too slower than openssh client

### ISSUE THREAD

**Issue by felipeaf** [mscdex/ssh2#207]
For me, the time for a complete upload is between 2 and 3 times than the same upload with openssh clients (either sftp or scp).

By sftp from openssh client:

    du -h VMware-Player-6.0.2-1744117.x86_64.bundle
    192M	VMware-Player-6.0.2-1744117.x86_64.bundle
    XXX@localhost's password: 
    Connected to localhost.
    sftp> cd /tmp
    sftp> put /home/XXX/VMware-Player-6.0.2-1744117.x86_64.bundle
    Uploading /home/XXX/VMware-Player-6.0.2-1744117.x86_64.bundle to /tmp/VMware-Player-6.0.2-1744117.x86_64.bundle
    /home/felipe.ferreira/VMware-Player-6.0.2-1744117.x86_64.bundle                                                            100%  191MB  47.8MB/s   00:04

(Above 4, sometimes 3 seconds!).
Now, my test program in node with ssh2 module:

```javascript
var Connection = require('ssh2');
var path = require("path");
var fs = require("fs");


/* CONFIG */
var file = "/home/XXX/Downloads/VMware-Player-6.0.2-1744117.x86_64.bundle";
var remoteFile= path.join('/tmp', path.basename(file));
var config = {
  host: 'localhost',
  username: process.env.USER,
  //privateKey: fs.readFileSync(path.join(process.env.HOME,'.ssh/id_rsa'))
  password: ""
};
/* END OF CONFIG PART */

var conn = new Connection();
conn.on('ready', function() {
  console.log('Connection :: ready');
  conn.sftp(function(err, sftp) {
    if(err) {
      console.log(err);
      process.exit(1);
    }
    var beginDate = new Date();
    console.log("Starting upload at", beginDate);
    sftp.fastPut(file, remoteFile, function(err) {
      var endDate = new Date();
      if(err)
        console.log("Upload has finished at ", endDate, "with error", err);
      else
        console.log("Upload has finished sucessfuly at", endDate);
      console.log("Duration:", (endDate - beginDate)/1000, "seconds");
      process.exit(0);
    });
  });
}).connect(config);
```

The output:

    node testssh2.js
    Connection :: ready
    Starting upload at Wed Dec 24 2014 11:42:54 GMT-0200 (BRST)
    Upload has finished sucessfuly at Wed Dec 24 2014 11:43:04 GMT-0200 (BRST)
    Duration: 9.916 seconds

Normally about 10, sometimes 12 seconds!

For now, i have to use child_process instead ssh2 because of lack of performance for my problem.

**Comment by jfenton** [comment]
Also seeing this (but for download):

21 megabyte file over SFTP via scp: 20.3MB/sec
21 megabyte file over SFTP via ssh2: 11.42MB/sec
(downloading 5 files simultaneously via ssh2 == 2.26MB/s per file)

Also tested on larger corpus of files with same speeds.

Using sftp.fastGet with concurrency 25 chunkSize 32k (twiddled, but this is best performance I could get)

**Comment by mscdex** [comment]
@jfenton Is `scp` symlinked to `sftp` on the system you're testing with? If not, can you test with `sftp` too?

**Comment by jfenton** [comment]
Negative, separate binaries, have tried with sftp - same performance.

**Comment by jfenton** [comment]
(copying in details from merged ticket)

As reported by others, I'm seeing a 0.5x performance of SFTP downloads using ssh2:

21 megabyte file over SFTP via scp: 20.3MB/sec
21 megabyte file over SFTP via ssh2: 11.42MB/sec
(downloading 5 files simultaneously via ssh2 == 2.26MB/s per file)

Also tested on larger corpus of files with same speeds.

Using sftp.fastGet with concurrency 25 chunkSize 32k (twiddled both, but this is best performance I could get)

sshd DEBUG3 output for SCP and SSH2 at https://gist.github.com/jfenton/fcc8cd0003a2a5e66d79

(one point of note is that SCP negotiates aes128, SSH2 negotiates aes256 but I'm not sure that would account for the performance decrease and can't find a way to change the cipher in ssh2).

Corpus of 49 files, 994MB

scp *
18s @ 20.3MB/s

ssh2 with 5 concurrent downloads
1m49s @ 2.96MB/s per file * 5 concurrent downloads = 14.8MB/s

**Comment by mscdex** [comment]
I just tested `fastGet()` on the rewrite branch and I consistently saw ~11MB/s faster transfer compared to the master branch. I then compared the rewrite branch results with that of `pscp` (which should hopefully be pretty close to OpenSSH's scp/sftp performance-wise) and it's about the same (maybe a tad faster than `pscp`). FWIW `pscp` was also using the aes-256-ctr.

Can you all test with the rewrite branch and let me know what you find?: `npm install https://github.com/mscdex/ssh2/tarball/rewrite`

**Comment by jfenton** [comment]
I will try with rewrite. In the meantime, here's a nice flamegraph showing where time is being spent.

https://gist.github.com/jfenton/87e64d6c5da5509e4c48
(copy it into a local .svg file and drop back into your browser)

**Comment by mscdex** [comment]
Yeah, the flamegraph makes sense for the current master branch. The rewrite uses a tweaked version of the old parser that is more efficient.

**Comment by jfenton** [comment]
Substantial improvement. With your rewrite, I'm seeing:

21.9MB/s for a single file
6.1MB/s per file with 5 concurrent downloads = 31MB/s

Nice work :)

When will this hit master?

**Comment by mscdex** [comment]
@jfenton I'm not sure yet. There's at least a few tests that need to be written yet (especially testing basic sftp subsystem starting and such).

Also I haven't decided how to handle backwards compatibility as far as sftp goes. The problem there is that since I have implemented sftp as a stream, I had to rename a couple of methods (namely `write()` to `writeData()` and `read()` to `readData()`) which obviously conflict with node's streams interface. Currently I'm just passing these streams as-is to the end user and this breaks backwards compatibility and `readData()` and `writeData()` are just not as nice as `read()` or `write()`. So I'm still debating whether I should create a silly proxy thing that encapsulates an sftp stream and has `read()` and `write()` methods that defer to `readData()` and `writeData()` on the stream.
